src/Workspaces/Remote/Core/Services/RemoteWorkspace.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using Microsoft.CodeAnalysis.Host;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// primary workspace for remote host. no one except solution service can update this workspace
+    /// </summary>
+    internal class RemoteWorkspace : Workspace
+    {
+        public const string WorkspaceKind_RemoteWorkspace = "RemoteWorkspace";
+
+        public RemoteWorkspace()
+            : base(RoslynServices.HostServices, workspaceKind: RemoteWorkspace.WorkspaceKind_RemoteWorkspace)
+        {
+            PrimaryWorkspace.Register(this);
+
+            Options = Options.WithChangedOption(CacheOptions.RecoverableTreeLengthThreshold, 0);
+        }
+
+        public override bool CanApplyChange(ApplyChangesKind feature)
+        {
+            // apply change is not allowed
+            return false;
```

Comment:
might be an issue if we have features that do light-up based on if they think they can change the workspace or not.